In [1]:
import sys;
sys.path.insert(0, '..')

## Chapter 4 Code Snippets and Listings

### Two-qubit states (section 4.2.1)

A two-qubit quantum state can be represented as a list of four complex numbers.
We can encode the general form of a two-qubit state using the Python list below:

In [2]:
from math import sqrt, cos, sin

[p0, p1, p2, p3] = [1, 0, 0, 0]
[theta0, theta1, theta2, theta3] = [0, 0, 0, 0]

state = [sqrt(p0) * (cos(theta0) + 1j * sin(theta0)),
         sqrt(p1) * (cos(theta1) + 1j * sin(theta1)),
         sqrt(p2) * (cos(theta2) + 1j * sin(theta2)),
         sqrt(p3) * (cos(theta3) + 1j * sin(theta3))]

We can use the built-in `random` package in Python to generate example probabilities and directions:

In [3]:
import random
from math import pi

random.seed(123456789) # chooses a seed so that we get reproducible results

probs = [random.random() for _ in range(4)] # generates 4 random numbers
total = sum(probs)
probs = [p/total for p in probs] # normalizes each amplitude so that the probabilities add up to 1

angles = [random.uniform(0, 2*pi) for _ in range(4)] # generates 4 random angles in radians

state = [sqrt(p)*(cos(a) + 1j*sin(a)) for (p, a) in zip(probs, angles)] # builds the quantum state list

In [4]:
from util import print_state_table

print_state_table(state)


Outcome  Binary  Amplitude           Magnitude  Direction  Amplitude Bar             Probability
------------------------------------------------------------------------------------------------
0        00      0.1743 - i0.4266    0.4609      -67.22°   ███████████               0.2124
1        01     -0.3383 + i0.2551    0.4237      142.98°   ██████████                0.1795
2        10      0.5293 - i0.2208    0.5735      -22.36°   █████████████             0.3289
3        11     -0.1973 - i0.4902    0.5284     -111.80°   ████████████              0.2792



**Product states: composing a state from two independent single-qubit states**

Shortcut `cis` function:

In [5]:
def cis(theta):
    return cos(theta) + 1j*sin(theta)

The first state has probability $p = 0.75$ for outcome 0, and directions $\theta_0 = 0^\circ$ and $\theta_1 = 60^\circ$:

In [6]:
p = 0.75
theta0 = 0
theta1 = 60/(180/pi) # converts to radians
first_state = [sqrt(p)*cis(theta0), sqrt(1-p)*cis(theta1)]
print([round(amp.real, 5)+1j*round(amp.imag, 5) for amp in first_state])

[(0.86603+0j), (0.25+0.43301j)]


The second state has $q = 0.5$ as the probability for outcome 0, and directions $\phi_0 = 0^\circ$ and $\phi_1 = -120^\circ$:

In [7]:
q = 0.5
phi0 = 0
phi1 = -120/(180/pi)
second_state = [sqrt(q)*cis(phi0), sqrt(1-q)*cis(phi1)]
print([round(amp.real, 5)+1j*round(amp.imag, 5) for amp in second_state])

[(0.70711+0j), (-0.35355-0.61237j)]


If we create a two-qubit state from these two single-qubit states, the new amplitudes will be all the possible products of the two pairs of amplitudes:

In [8]:
new_state = [first_state[0]*second_state[0], first_state[0]*second_state[1],
             first_state[1]*second_state[0], first_state[1]*second_state[1]]
print([round(amp.real, 5)+1j*round(amp.imag, 5) for amp in new_state])

[(0.61237+0j), (-0.30619-0.53033j), (0.17678+0.30619j), (0.17678-0.30619j)]


Alternate definition:

In [9]:
new_state = [sqrt(p*q)*cis(theta0 + phi0), sqrt(p*(1-q))*cis(theta0 + phi1),
            sqrt((1-p)*q)*cis(theta1 + phi0), sqrt((1-p)*(1-q))*cis(theta1 + phi1)]
print([round(amp.real, 5)+1j*round(amp.imag, 5) for amp in new_state])

[(0.61237+0j), (-0.30619-0.53033j), (0.17678+0.30619j), (0.17678-0.30619j)]


**Examples of non-product states: Bell states**

In [10]:
bell_state1 = [sqrt(0.5), 0.0, 0.0, sqrt(0.5)]

bell_state2 = [sqrt(0.5), 0.0, 0.0, -sqrt(0.5)]

In [11]:
print_state_table(bell_state1)


Outcome  Binary  Amplitude           Magnitude  Direction  Amplitude Bar             Probability
------------------------------------------------------------------------------------------------
0        00      0.7071 + i0.0000    0.7071        0.00°   ████████████████          0.5   
1        01      0.0000 + i0.0000    0.0                                             0.0   
2        10      0.0000 + i0.0000    0.0                                             0.0   
3        11      0.7071 + i0.0000    0.7071        0.00°   ████████████████          0.5   



In [12]:
print_state_table(bell_state2)


Outcome  Binary  Amplitude           Magnitude  Direction  Amplitude Bar             Probability
------------------------------------------------------------------------------------------------
0        00      0.7071 + i0.0000    0.7071        0.00°   ████████████████          0.5   
1        01      0.0000 + i0.0000    0.0                                             0.0   
2        10      0.0000 + i0.0000    0.0                                             0.0   
3        11     -0.7071 + i0.0000    0.7071      180.00°   ████████████████          0.5   



In [13]:
bell_state3 = [0.0, sqrt(0.5), sqrt(0.5), 0.0]

bell_state4 = [0.0, sqrt(0.5), -sqrt(0.5), 0.0]

In [14]:
print_state_table(bell_state3)


Outcome  Binary  Amplitude           Magnitude  Direction  Amplitude Bar             Probability
------------------------------------------------------------------------------------------------
0        00      0.0000 + i0.0000    0.0                                             0.0   
1        01      0.7071 + i0.0000    0.7071        0.00°   ████████████████          0.5   
2        10      0.7071 + i0.0000    0.7071        0.00°   ████████████████          0.5   
3        11      0.0000 + i0.0000    0.0                                             0.0   



In [15]:
print_state_table(bell_state4)


Outcome  Binary  Amplitude           Magnitude  Direction  Amplitude Bar             Probability
------------------------------------------------------------------------------------------------
0        00      0.0000 + i0.0000    0.0                                             0.0   
1        01      0.7071 + i0.0000    0.7071        0.00°   ████████████████          0.5   
2        10     -0.7071 + i0.0000    0.7071      180.00°   ████████████████          0.5   
3        11      0.0000 + i0.0000    0.0                                             0.0   



### Multi-qubit states (section 4.2.2)

Let's define a state with the following eight complex numbers:

In [16]:
amplitude_list = [(0.09858+0.03637j),
                (0.07478+0.06912j),
                (0.04852+0.10526j),
                (0.00641+0.16322j),
                (-0.12895+0.34953j),
                (0.58403-0.6318j),
                (0.18795-0.08665j),
                (0.12867-0.00506j)]

Listing 4.1 Function to check whether a list is a valid quantum state

In [17]:
from math import log2, ceil, floor
from util import is_close

def is_power_of_two(m):
    return ceil(log2(m)) == floor(log2(m))

def prepare_state(*a):
    state = [a[k] for k in range(len(a))]
    assert (is_power_of_two(len(state))) # checks that the length of the list is a power of 2
    assert (is_close(sum([abs(state[k]) ** 2 for k in range(len(state))]), 1.0)) # checks that the squared magnitudes add up to 1
    return state # if the conditions are met, we return the state

In [18]:
state = prepare_state(*amplitude_list)

**Building state tables with Python list comprehensions**

List comprehension for outcomes and corresponding probabilities:

In [19]:
print([[k, state[k]] for k in range(len(state))])

[[0, (0.09858+0.03637j)], [1, (0.07478+0.06912j)], [2, (0.04852+0.10526j)], [3, (0.00641+0.16322j)], [4, (-0.12895+0.34953j)], [5, (0.58403-0.6318j)], [6, (0.18795-0.08665j)], [7, (0.12867-0.00506j)]]


We can use the following list comprehension to add probabilities with direction, derived from amplitudes:

In [20]:
from math import atan2

table1 = [
    [
        k,
        round(atan2(state[k].imag, state[k].real) / (2 * pi) * 360, 5),
        round(abs(state[k]) ** 2, 5)
    ]
    for k in range(len(state))
]

for row in table1:
    print(row)

[0, 20.25098, 0.01104]
[1, 42.74755, 0.01037]
[2, 65.25248, 0.01343]
[3, 87.75103, 0.02668]
[4, 110.25023, 0.1388]
[5, -47.25, 0.74026]
[6, -24.75097, 0.04283]
[7, -2.25202, 0.01658]


We can build an expanded version of the state table that includes the direction and magnitude of amplitudes, as well as the probability of the outcomes:

In [21]:
expanded_table = [
    [
        k,
        state[k],
        round(atan2(state[k].imag, state[k].real) / (2 * pi) * 360, 5),
        round(abs(state[k]), 5),
        round(abs(state[k]) ** 2, 5)
    ]
    for k in range(len(state))
]

for row in expanded_table:
    print(row)

[0, (0.09858+0.03637j), 20.25098, 0.10508, 0.01104]
[1, (0.07478+0.06912j), 42.74755, 0.10183, 0.01037]
[2, (0.04852+0.10526j), 65.25248, 0.1159, 0.01343]
[3, (0.00641+0.16322j), 87.75103, 0.16335, 0.02668]
[4, (-0.12895+0.34953j), 110.25023, 0.37256, 0.1388]
[5, (0.58403-0.6318j), -47.25, 0.86038, 0.74026]
[6, (0.18795-0.08665j), -24.75097, 0.20696, 0.04283]
[7, (0.12867-0.00506j), -2.25202, 0.12877, 0.01658]


we can use the following list comprehension to get amplitudes from directions and probabilities:

In [22]:
table2 = [
    [
        row[0],
        (
            round(sqrt(row[2]) * cos(row[1] / (180 / pi)), 5) +
            round(sqrt(row[2]) * sin(row[1] / (180 / pi)), 5) * 1j
        )
    ]
    for row in table1
]

for row in table2:
        print(row)

[0, (0.09858+0.03637j)]
[1, (0.07478+0.06912j)]
[2, (0.04851+0.10524j)]
[3, (0.00641+0.16321j)]
[4, (-0.12895+0.34953j)]
[5, (0.58403-0.6318j)]
[6, (0.18794-0.08665j)]
[7, (0.12866-0.00506j)]


In [23]:
print_state_table(state)


Outcome  Binary  Amplitude           Magnitude  Direction  Amplitude Bar             Probability
------------------------------------------------------------------------------------------------
0        000     0.0986 + i0.0364    0.1051       20.26°   ██                        0.011 
1        001     0.0748 + i0.0691    0.1018       42.73°   ██                        0.0104
2        010     0.0485 + i0.1053    0.1159       65.27°   ██                        0.0134
3        011     0.0064 + i0.1632    0.1633       87.75°   ███                       0.0267
4        100    -0.1290 + i0.3495    0.3726      110.26°   ████████                  0.1388
5        101     0.5840 - i0.6318    0.8604      -47.75°   ████████████████████      0.7403
6        110     0.1880 - i0.0867    0.207       -24.24°   ████                      0.0428
7        111     0.1287 - i0.0051    0.1288       -2.73°   ███                       0.0166



### Simulating multi-qubit states in Python (section 4.2.3)

Listing 4.2 Function to create a default quantum state

In [24]:
def init_state(n):
    state = [0 for _ in range(2 ** n)] # given n qubits, the state will contain 2^n complex numbers
    state[0] = 1 # the amplitude corresponding to outcome 0 (the first amplitude in the list) will have a value of 1
    return state

Initialize a two-qubit state:

In [25]:
state = init_state(2)
state

[1, 0, 0, 0]

### Pair selection in Python (section 4.3.2)

Listing 4.3 Traverse-by-chunk method for selecting pairs

In [26]:
def pair_generator_pattern(n, t):
    distance = int(2 ** t) # distance is the size of each chunk

    for j in range(2**(n-t-1)):
        for k0 in range(2*j*distance, (2*j+1)*distance): # get the 0 side of each pair
            k1 = k0 + distance # get the 1 side of the pair by adding the distance (2^t)
            yield k0, k1

Example of generate the pairs using this method for three qubits (`n = 3`) and target qubit 0 (`t = 1`):

In [27]:
for (k0, k1) in pair_generator_pattern(3, 0):
    print(k0, k1)

0 1
2 3
4 5
6 7


### Simulating amplitude changes (section 4.3.3)

Listing 4.4 Functions for simulating a gate transformation on a multi-qubit state

In [28]:
def process_pair(state, gate, k0, k1):
    x = state[k0] # get the original amplitudes of the pair
    y = state[k1] # computes the amplitudes given the gate definition and replace the old amplitudes in the state list
    state[k0] = x * gate[0][0] + y * gate[0][1]
    state[k1] = x * gate[1][0] + y * gate[1][1]

def transform(state, t, gate):
    n = int(log2(len(state))) # gets the number of qubits in the state
    for (k0, k1) in pair_generator(n, t): # calls pair_generator which returns pairs as tuples given a number of qubits (n) and target qubit (t)
        process_pair(state, gate, k0, k1) # for each pair, call process_pair to compute the new amplitudes

In [29]:
pair_generator = pair_generator_pattern

Example state:

In [30]:
state = [(0.09858+0.03637j),
        (0.07478+0.06912j),
        (0.04852+0.10526j),
        (0.00641+0.16322j),
        (-0.12895+0.34953j),
        (0.58403-0.6318j),
        (0.18795-0.08665j),
        (0.12867-0.00506j)]

Apply an X-gate to target qubit 0:

In [31]:
from sim_gates import * # imports the gate definitions added to the source code from chapter 3

transform(state, 0, x)
state

[(0.07478+0.06912j),
 (0.09858+0.03637j),
 (0.00641+0.16322j),
 (0.04852+0.10526j),
 (0.58403-0.6318j),
 (-0.12895+0.34953j),
 (0.12867-0.00506j),
 (0.18795-0.08665j)]

### Encoding a uniform distribution in a multi-qubit quantum system (section 4.3.4)

First, we initialize a three-qubit state:

In [32]:
state = init_state(3)

In [33]:
print(state)

[1, 0, 0, 0, 0, 0, 0, 0]


Apply a Hadamard gate to target qubit 0:

In [34]:
transform(state, 0, h)
print(state)

[0.7071067811865475, 0.7071067811865475, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]


Apply a Hadamard gate to target qubit 1:

In [35]:
state = init_state(3)
transform(state, 1, h)
print(state)

[0.7071067811865475, 0.0, 0.7071067811865475, 0.0, 0.0, 0.0, 0.0, 0.0]


Apply a Hadamard gate to each qubit:

In [36]:
state = init_state(3)
transform(state, 0, h)
transform(state, 1, h)
transform(state, 2, h)
print(state)

[0.3535533905932737, 0.3535533905932737, 0.3535533905932737, 0.3535533905932737, 0.3535533905932737, 0.3535533905932737, 0.3535533905932737, 0.3535533905932737]


Applying a Hadamard gate to each qubit creates a state where each outcome has an equal probability.
Thus, the state will have a uniform probability distribution.

In [37]:
print_state_table(state)


Outcome  Binary  Amplitude           Magnitude  Direction  Amplitude Bar             Probability
------------------------------------------------------------------------------------------------
0        000     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
1        001     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
2        010     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
3        011     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
4        100     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
5        101     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
6        110     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
7        111     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 



### Simulating controlled gates in Python (section 4.4.1)

Listing 4.5 Applying controlled gate transformations to a state

In [38]:
def is_bit_set(m, k):
    return m & (1 << k) != 0

def c_transform(state, c, t, gate):
    n = int(log2(len(state))) # uses the same pair generation function as for regular transforms with a filter to check if the control position is 1
    for (k0, k1) in filter(lambda p: is_bit_set(p[0], c), pair_generator(n, t)): # once we get the right subset of pairs, we recombine the amplitudes according to the gate
        process_pair(state, gate, k0, k1)

Example state:

In [39]:
state = [(0.09858+0.03637j),
        (0.07478+0.06912j),
        (0.04852+0.10526j),
        (0.00641+0.16322j),
        (-0.12895+0.34953j),
        (0.58403-0.6318j),
        (0.18795-0.08665j),
        (0.12867-0.00506j)]

Apply a controlled X-gate to target qubit 2 with control qubit 1:

In [40]:
c_transform(state, 1, 2, x)
print(state)

[(0.09858+0.03637j), (0.07478+0.06912j), (0.18795-0.08665j), (0.12867-0.00506j), (-0.12895+0.34953j), (0.58403-0.6318j), (0.04852+0.10526j), (0.00641+0.16322j)]


### Simulating multi-control gates in Python (section 4.4.2)

Listing 4.6 Applying multicontrolled gate transformations to a state

In [41]:
def mc_transform(state, cs, t, gate):
    assert not t in cs # the target qubit cannot be the same as the control qubits
    n = int(log2(len(state))) # checks that the pairs have 1 in all the control qubit positions
    for (k0, k1) in filter(lambda p: all([is_bit_set(p[0], c) for c in cs]), pair_generator(n, t)):
        process_pair(state, gate, k0, k1) # recombines the amplitudes of the selected pairs

Example state:

In [42]:
state = [(0.09858+0.03637j), (0.07478+0.06912j), (0.04852+0.10526j),
         (0.00641+0.16322j), (-0.12895+0.34953j), (0.58403-0.6318j),
         (0.18795-0.08665j), (0.12867-0.00506j)]

If we apply a controlled transformation with two control qubits to a three-qubit state, there will be only one pair.
For example, if the target qubit is 0, and the control qubits are 1 and 2, the pair selected will be '110' and '111' (outcomes 6 and 7).


In [43]:
mc_transform(state, [1, 2], 0, x)
print(state)

[(0.09858+0.03637j), (0.07478+0.06912j), (0.04852+0.10526j), (0.00641+0.16322j), (-0.12895+0.34953j), (0.58403-0.6318j), (0.12867-0.00506j), (0.18795-0.08665j)]


### Simulating measurement of multi-qubit states (section 4.5.1)

Listing 4.7 Simulating measurement of a quantum state

In [44]:
from random import choices
from collections import Counter

def measure(state, shots):
    samples = choices(
       range(len(state)),
       [abs(state[k])**2 for k in range(len(state))],
       k=shots)
    counts = {}
    for (k, v) in Counter(samples).items():
        counts[k] = v
    return counts

Example state:

In [45]:
state = [(0.09858+0.03637j), (0.07478+0.06912j), (0.04852+0.10526j),
         (0.00641+0.16322j), (-0.12895+0.34953j), (0.58403-0.6318j),
         (0.18795-0.08665j), (0.12867-0.00506j)]

In [46]:
probabilities = [[k, abs(state[k])**2] for k in range(len(state))]

for i in probabilities:
    print("probability of outcome", i[0], ": ", round(i[1], 3))

probability of outcome 0 :  0.011
probability of outcome 1 :  0.01
probability of outcome 2 :  0.013
probability of outcome 3 :  0.027
probability of outcome 4 :  0.139
probability of outcome 5 :  0.74
probability of outcome 6 :  0.043
probability of outcome 7 :  0.017


Let's simulate the outcomes of 100 executions of the computation which creates our example state:

In [47]:
samples = measure(state, 100)
print(samples)

{5: 83, 2: 3, 4: 10, 3: 2, 1: 1, 6: 1}


### Quantum registers and circuits in code (section 4.5.2)

Listing 4.8 `QuantumRegister` class

In [48]:
class QuantumRegister:
    def __init__(self, size, shift=0):
        self.size = size
        self.shift = shift

    def __getitem__(self, key):
        if isinstance(key, slice):
            return [self[ii] for ii in range(*key.indices(len(self)))]
        elif isinstance(key, int):
            if key < 0:
                key += len(self)
            assert(0 <= key < self.size)
            return self.shift + key

    def __len__(self):
        return self.size

    def __iter__(self):
        return list([self.shift + i for i in range(self.size)])

    def __reversed__(self):
        return list([self.shift + i for i in range(self.size)[::-1]])

Listing 4.9 `QuantumTransformation` class

In [49]:
class QuantumTransformation:
    def __init__(self, gate, target, controls=[], name=None, arg=None):
        self.gate = gate
        self.target= target
        self.controls = controls
        self.name = name
        self.arg = arg

    def __str__(self):
        return rf'{self.name} {round(self.arg, 2) if self.arg is not None else ""} '\
               f'{self.controls} {self.target}'

    def __copy__(self):
        return QuantumTransformation(
            self.gate,
            self.target,
            self.controls,
            self.name,
            self.arg
        )

Listing 4.10 Partial implementation of the `QuantumCircuit` class

In [50]:
class QuantumCircuit:
    def __init__(self, *args):
        bits = 0
        regs = []
        for register in args:
            register.shift = bits
            bits += register.size
            regs.append(register.size)

        self.state = init_state(bits)
        self.transformations = []
        self.regs = regs
        self.reports = {}

    def initialize(self, state):
        self.state = state

    def x(self, t):
        self.transformations.append(
           QuantumTransformation(x, t, [], 'x'))

    def h(self, t):
        self.transformations.append(
           QuantumTransformation(h, t, [], 'h'))

    def ry(self, theta, t):
        self.transformations.append(
           QuantumTransformation(ry(theta), t, [], 'ry', theta))

    def cx(self, c, t):
        self.transformations.append(
           QuantumTransformation(x, t, [c], 'x'))

    def mcx(self, cs, t):
        self.transformations.append(
           QuantumTransformation(x, t, cs, 'x'))

    def measure(self, shots=0):
        state = self.run()
        samples = measure(state, shots)
        return {'state vector': state, 'counts': samples}

    def run(self):
        for tr in self.transformations:
            cs = tr.controls
            if len(cs) == 0:
                transform(self.state, tr.target, tr.gate)
            elif len(cs) == 1:
                c_transform(self.state, cs[0], tr.target, tr.gate)
            else:
                mc_transform(self.state, cs, tr.target, tr.gate)
        self.transformations = []
        return self.state

Example three-qubit circuit:

In [51]:
q = QuantumRegister(3)
qc = QuantumCircuit(q)

qc.h(q[0])
qc.h(q[1])
qc.mcx([q[0], q[1]], q[2])

When we call the `run` function, the respective transformations will be applied to the state object, and the transformed state is returned:

In [52]:
state = qc.run()

In [53]:
print_state_table(state)


Outcome  Binary  Amplitude           Magnitude  Direction  Amplitude Bar             Probability
------------------------------------------------------------------------------------------------
0        000     0.5000 + i0.0000    0.5           0.00°   ███████████               0.25  
1        001     0.5000 + i0.0000    0.5           0.00°   ███████████               0.25  
2        010     0.5000 + i0.0000    0.5           0.00°   ███████████               0.25  
3        011     0.0000 + i0.0000    0.0                                             0.0   
4        100     0.0000 + i0.0000    0.0                                             0.0   
5        101     0.0000 + i0.0000    0.0                                             0.0   
6        110     0.0000 + i0.0000    0.0                                             0.0   
7        111     0.5000 + i0.0000    0.5           0.00°   ███████████               0.25  



We can simulate measurement on the resulting state with the `measure` function:

In [54]:
samples = measure(state, 1000)
print(samples)

{2: 245, 7: 238, 0: 267, 1: 250}


### Reimplementing the uniform distribution with registers and circuits (section 4.5.3)

In [55]:
q = QuantumRegister(3)
qc = QuantumCircuit(q)

for i in range(len(q)):
    qc.h(q[i])

state = qc.run()

In [56]:
print_state_table(state)


Outcome  Binary  Amplitude           Magnitude  Direction  Amplitude Bar             Probability
------------------------------------------------------------------------------------------------
0        000     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
1        001     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
2        010     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
3        011     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
4        100     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
5        101     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
6        110     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 
7        111     0.3536 + i0.0000    0.3536        0.00°   ████████                  0.125 



Listing 4.11 Encoding the uniform distribution in a quantum state

In [57]:
def uniform(n):
    q = QuantumRegister(n)
    qc = QuantumCircuit(q)

    for i in range(len(q)):
        qc.h(q[i])

    return qc

### Encoding the Binomial distribution in a multi-qubit state (section 4.5.4)

In [58]:
q = QuantumRegister(3)
qc = QuantumCircuit(q)

for i in range(len(q)):
    qc.ry(pi/3, q[i])

state = qc.run()

In [59]:
print_state_table(state)


Outcome  Binary  Amplitude           Magnitude  Direction  Amplitude Bar             Probability
------------------------------------------------------------------------------------------------
0        000     0.6495 + i0.0000    0.6495        0.00°   ███████████████           0.4219
1        001     0.3750 + i0.0000    0.375         0.00°   █████████                 0.1406
2        010     0.3750 + i0.0000    0.375         0.00°   █████████                 0.1406
3        011     0.2165 + i0.0000    0.2165        0.00°   █████                     0.0469
4        100     0.3750 + i0.0000    0.375         0.00°   █████████                 0.1406
5        101     0.2165 + i0.0000    0.2165        0.00°   █████                     0.0469
6        110     0.2165 + i0.0000    0.2165        0.00°   █████                     0.0469
7        111     0.1250 + i0.0000    0.125         0.00°   ██                        0.0156



Listing 4.12 Encoding the binomial distribution in a quantum state

In [60]:
def binomial(n, theta):
    q = QuantumRegister(n)
    qc = QuantumCircuit(q)

    for i in range(len(q)): # Iterate through each qubit in the register and apply an Ry-gate
        qc.ry(theta, q[i])
    
    return qc

### Implementing the Bell states (section 4.5.5)

First Bell state:

In [61]:
q = QuantumRegister(2)
qc = QuantumCircuit(q)

qc.h(q[0])
qc.cx(q[0], q[1])

state = qc.run()

In [62]:
print_state_table(state)


Outcome  Binary  Amplitude           Magnitude  Direction  Amplitude Bar             Probability
------------------------------------------------------------------------------------------------
0        00      0.7071 + i0.0000    0.7071        0.00°   ████████████████          0.5   
1        01      0.0000 + i0.0000    0.0                                             0.0   
2        10      0.0000 + i0.0000    0.0                                             0.0   
3        11      0.7071 + i0.0000    0.7071        0.00°   ████████████████          0.5   



Third Bell state:

In [63]:
q = QuantumRegister(2)
qc = QuantumCircuit(q)

qc.h(q[0])
qc.x(q[1])
qc.cx(q[0], q[1])

state = qc.run()

In [64]:
print_state_table(state)


Outcome  Binary  Amplitude           Magnitude  Direction  Amplitude Bar             Probability
------------------------------------------------------------------------------------------------
0        00      0.0000 + i0.0000    0.0                                             0.0   
1        01      0.7071 + i0.0000    0.7071        0.00°   ████████████████          0.5   
2        10      0.7071 + i0.0000    0.7071        0.00°   ████████████████          0.5   
3        11      0.0000 + i0.0000    0.0                                             0.0   

